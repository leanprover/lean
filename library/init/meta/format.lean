/-
Copyright (c) 2016 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Leonardo de Moura
-/
prelude
import init.meta.options

inductive format.color
| red | green | orange | blue | pink | cyan | grey

meta_constant format : Type₁
meta_constant format.line          : format
meta_constant format.space         : format
meta_constant format.nil           : format
meta_constant format.compose       : format → format → format
meta_constant format.nest          : nat → format → format
meta_constant format.highlight     : format → color → format
meta_constant format.group         : format → format
meta_constant format.of_string     : string → format
meta_constant format.of_nat        : nat → format
meta_constant format.flatten       : format → format
meta_constant format.to_string     : format → options → string
meta_constant format.of_options    : options → format
meta_constant format.is_nil        : format → bool
meta_constant trace_fmt {A : Type} : format → (unit → A) → A


attribute [instance]
meta_definition format.is_inhabited : inhabited format :=
inhabited.mk format.space

attribute [instance]
meta_definition format_has_append : has_append format :=
has_append.mk format.compose

attribute [instance]
meta_definition format_has_to_string : has_to_string format :=
has_to_string.mk (λ f, format.to_string f options.mk)

structure [class] has_to_format (A : Type) :=
(to_format : A → format)

attribute [instance]
meta_definition format_has_to_format : has_to_format format :=
has_to_format.mk id

meta_definition to_fmt {A : Type} [has_to_format A] : A → format :=
has_to_format.to_format

attribute [instance]
meta_definition coe_nat_to_format : has_coe nat format :=
has_coe.mk format.of_nat

attribute [instance]
meta_definition coe_string_to_format : has_coe string format :=
has_coe.mk format.of_string

open format list

meta_definition format.when {A : Type} [has_to_format A] : bool → A → format
| tt a := to_fmt a
| ff a := nil

attribute [instance]
meta_definition options.has_to_format : has_to_format options :=
has_to_format.mk (λ o, format.of_options o)

attribute [instance]
meta_definition bool.has_to_format : has_to_format bool :=
has_to_format.mk (λ b, if b = tt then of_string "tt" else of_string "ff")

attribute [instance]
meta_definition decidable.has_to_format {p : Prop} : has_to_format (decidable p) :=
has_to_format.mk (λ b, if p then of_string "tt" else of_string "ff")

attribute [instance]
meta_definition string.has_to_format : has_to_format string :=
has_to_format.mk (λ s, format.of_string s)

attribute [instance]
meta_definition nat.has_to_format : has_to_format nat :=
has_to_format.mk (λ n, format.of_nat n)

attribute [instance]
meta_definition char.has_to_format : has_to_format char :=
has_to_format.mk (λ c : char, format.of_string [c])

meta_definition list.to_format_aux {A : Type} [has_to_format A] : bool → list A → format
| _  []      := to_fmt ""
| tt (x::xs) := to_fmt x ++ list.to_format_aux ff xs
| ff (x::xs) := to_fmt "," ++ line ++ to_fmt x ++ list.to_format_aux ff xs

meta_definition list.to_format {A : Type} [has_to_format A] : list A → format
| []      := to_fmt "[]"
| (x::xs) := to_fmt "[" ++ group (nest 1 (list.to_format_aux tt (x::xs))) ++ to_fmt "]"

attribute [instance]
meta_definition list.has_to_format {A : Type} [has_to_format A] : has_to_format (list A) :=
has_to_format.mk list.to_format

attribute [instance] string.has_to_format

attribute [instance]
meta_definition name.has_to_format : has_to_format name :=
has_to_format.mk (λ n, to_fmt (to_string n))

attribute [instance]
meta_definition unit.has_to_format : has_to_format unit :=
has_to_format.mk (λ u, to_fmt "()")

attribute [instance]
meta_definition option.has_to_format {A : Type} [has_to_format A] : has_to_format (option A) :=
has_to_format.mk (λ o, option.cases_on o
  (to_fmt "none")
  (λ a, to_fmt "(some " ++ nest 6 (to_fmt a) ++ to_fmt ")"))

attribute [instance]
meta_definition sum.has_to_format {A B : Type} [has_to_format A] [has_to_format B] : has_to_format (sum A B) :=
has_to_format.mk (λ s, sum.cases_on s
  (λ a, to_fmt "(inl " ++ nest 5 (to_fmt a) ++ to_fmt ")")
  (λ b, to_fmt "(inr " ++ nest 5 (to_fmt b) ++ to_fmt ")"))

open prod

attribute [instance]
meta_definition prod.has_to_format {A B : Type} [has_to_format A] [has_to_format B] : has_to_format (prod A B) :=
has_to_format.mk (λ p, group (nest 1 (to_fmt "(" ++ to_fmt (pr1 p) ++ to_fmt "," ++ line ++ to_fmt (pr2 p) ++ to_fmt ")")))

open sigma

attribute [instance]
meta_definition sigma.has_to_format {A : Type} {B : A → Type} [has_to_format A] [s : ∀ x, has_to_format (B x)]
                                          : has_to_format (sigma B) :=
has_to_format.mk (λ p, group (nest 1 (to_fmt "⟨"  ++ to_fmt (pr1 p) ++ to_fmt "," ++ line ++ to_fmt (pr2 p) ++ to_fmt "⟩")))

open subtype

attribute [instance]
meta_definition subtype.has_to_format {A : Type} {P : A → Prop} [has_to_format A] : has_to_format (subtype P) :=
has_to_format.mk (λ s, to_fmt (elt_of s))

meta_definition format.bracket : string → string → format → format
| o c f := to_fmt o ++ nest (utf8_length o) f ++ to_fmt c

meta_definition format.paren (f : format) : format :=
format.bracket "(" ")" f

meta_definition format.cbrace (f : format) : format :=
format.bracket "{" "}" f

meta_definition format.sbracket (f : format) : format :=
format.bracket "[" "]" f

meta_definition format.dcbrace (f : format) : format :=
to_fmt "⦃" ++ nest 1 f ++ to_fmt "⦄"
