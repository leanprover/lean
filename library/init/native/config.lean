/-
Copyright (c) 2016 Jared Roesch. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Jared Roesch
-/
prelude
import init.data.bool.basic

namespace native

inductive backend
| CPP
| JS

-- eventually expose all the options here
-- if you change this you change the corresponding
-- code in library/native_compiler/native_compiler.cpp
record config :=
  (debug : bool)
  (compiler_backend : option backend)

end native
