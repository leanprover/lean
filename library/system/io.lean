/-
Copyright (c) 2017 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Luke Nelson, Jared Roesch and Leonardo de Moura
-/
import data.buffer

inductive io.error
| other     : string → io.error
| sys       : nat → io.error

structure io.terminal (m : Type → Type → Type) :=
(put_str     : string → m io.error unit)
(get_line    : m io.error string)
(cmdline_args : list string)

inductive io.mode
| read | write | read_write | append

structure io.file_system (handle : Type) (m : Type → Type → Type) :=
/- Remark: in Haskell, they also provide  (Maybe TextEncoding) and  NewlineMode -/
(mk_file_handle : string → io.mode → bool → m io.error handle)
(is_eof         : handle → m io.error bool)
(flush          : handle → m io.error unit)
(close          : handle → m io.error unit)
(read           : handle → nat → m io.error char_buffer)
(write          : handle → char_buffer → m io.error unit)
(get_line       : handle → m io.error char_buffer)
(stdin          : m io.error handle)
(stdout         : m io.error handle)
(stderr         : m io.error handle)

inductive io.process.stdio
| piped
| inherit
| null

structure io.process.spawn_args :=
/- Command name. -/
(cmd : string)
/- Arguments for the process -/
(args : list string := [])
/- Configuration for the process' stdin handle. -/
(stdin := stdio.inherit)
/- Configuration for the process' stdout handle. -/
(stdout := stdio.inherit)
/- Configuration for the process' stderr handle. -/
(stderr := stdio.inherit)
/- Working directory for the process. -/
(cwd : option string := none)

structure io.process (handle : Type) (m : Type → Type → Type) :=
(child : Type) (stdin : child → handle) (stdout : child → handle) (stderr : child → handle)
(spawn : io.process.spawn_args → m io.error child)
(wait : child → m io.error nat)

class io.interface :=
(m        : Type → Type → Type)
(monad    : Π e, monad (m e))
(catch    : Π e₁ e₂ α, m e₁ α → (e₁ → m e₂ α) → m e₂ α)
(fail     : Π e α, e → m e α)
(iterate  : Π e α, α → (α → m e (option α)) → m e α)
-- Primitive Types
(handle   : Type)
-- Interface Extensions
(term     : io.terminal m)
(fs       : io.file_system handle m)
(process  : io.process handle m)

variable [io.interface]

def io_core (e : Type) (α : Type) :=
io.interface.m e α

@[reducible] def io (α : Type) :=
io_core io.error α

instance io_core_is_monad (e : Type) : monad (io_core e) :=
io.interface.monad e

protected def io.fail {α : Type} (s : string) : io α :=
io.interface.fail io.error α (io.error.other s)

instance : monad_fail io :=
{ io_core_is_monad io.error with
  fail := @io.fail _ }

namespace io
def iterate {e α} (a : α) (f : α → io_core e (option α)) : io_core e α :=
interface.iterate e α a f

def forever {e} (a : io_core e unit) : io_core e unit :=
iterate () $ λ _, a >> return (some ())

def catch {e₁ e₂ α} (a : io_core e₁ α) (b : e₁ → io_core e₂ α) : io_core e₂ α :=
interface.catch e₁ e₂ α a b

instance : alternative io :=
{ interface.monad _ with
  orelse := λ _ a b, catch a (λ _, b),
  failure := λ _, io.fail "failure" }

def put_str : string → io unit :=
interface.term.put_str

def put_str_ln (s : string) : io unit :=
put_str s >> put_str "\n"

def get_line : io string :=
interface.term.get_line

def cmdline_args : io (list string) :=
return interface.term.cmdline_args

def print {α} [has_to_string α] (s : α) : io unit :=
put_str ∘ to_string $ s

def print_ln {α} [has_to_string α] (s : α) : io unit :=
print s >> put_str "\n"

def handle : Type :=
interface.handle

def mk_file_handle (s : string) (m : mode) (bin : bool := ff) : io handle :=
interface.fs.mk_file_handle s m bin

def stdin : io handle :=
interface.fs.stdin

def stderr : io handle :=
interface.fs.stderr

def stdout : io handle :=
interface.fs.stdout

namespace fs
def is_eof : handle → io bool :=
interface.fs.is_eof

def flush : handle → io unit :=
interface.fs.flush

def close : handle → io unit :=
interface.fs.close

def read : handle → nat → io char_buffer :=
interface.fs.read

def write : handle → char_buffer → io unit :=
interface.fs.write

def get_char (h : handle) : io char :=
do b ← read h 1,
   if h : b.size = 1 then return $ b.read ⟨0, h.symm ▸ zero_lt_one⟩
   else io.fail "get_char failed"

def get_line : handle → io char_buffer :=
interface.fs.get_line

def put_char (h : handle) (c : char) : io unit :=
write h (mk_buffer.push_back c)

def put_str (h : handle) (s : string) : io unit :=
write h (mk_buffer.append_string s)

def put_str_ln (h : handle) (s : string) : io unit :=
put_str h s >> put_str h "\n"

def read_to_end (h : handle) : io char_buffer :=
   iterate mk_buffer $ λ r,
     do done ← is_eof h,
     if done then return none
     else do
       c ← read h 1024,
       return $ some (r ++ c)

def read_file (s : string) (bin := ff) : io char_buffer :=
do h ← mk_file_handle s io.mode.read bin,
   read_to_end h

end fs

namespace proc
def child : Type := interface.process.child
def child.stdin : child → handle := interface.process.stdin
def child.stdout : child → handle := interface.process.stdout
def child.stderr : child → handle := interface.process.stderr
def spawn (p : io.process.spawn_args) : io child := interface.process.spawn p
def wait (c : child) : io nat := interface.process.wait c
end proc

end io

meta constant format.print_using [io.interface] : format → options → io unit

meta definition format.print (fmt : format) : io unit :=
format.print_using fmt options.mk

meta definition pp_using {α : Type} [has_to_format α] (a : α) (o : options) : io unit :=
format.print_using (to_fmt a) o

meta definition pp {α : Type} [has_to_format α] (a : α) : io unit :=
format.print (to_fmt a)

/-- Run the external process named by `cmd`, supplied with the arguments `args`.

    The process will run to completion with its output captured by a pipe, and
    read into `string` which is then returned.
-/
def io.cmd [io.interface] (cmd : string) (args : list string) : io string :=
do child ← io.proc.spawn {
    cmd := cmd,
    args := args,
    stdout := io.process.stdio.piped
  },
  buf ← io.fs.read child.stdout 1024,
  exitv ← io.proc.wait child,
  when (exitv ≠ 0) $ io.fail $ "process exited with status " ++ exitv.to_string,
  return buf.to_string

/-- Lift a monadic `io` action into the `tactic` monad. -/
meta constant tactic.run_io {α : Type} : (Π ioi : io.interface, @io ioi α) → tactic α
