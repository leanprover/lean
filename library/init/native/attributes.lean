prelude

import init.meta.tactic
import init.meta.constructor_tactic
import init.meta.attribute

definition ir_def : user_attribute := {
  name := `ir_def,
  descr := "makes the IR definition availible to the native compiler"
}

definition ir_decl : user_attribute := {
  name := `ir_decl,
  descr := "makes the IR declaration availible to the native compiler"
}

definition compiler_plugin : user_attribute := {
  name := `compiler_plugin,
  descr := "registers the compiler plugin for use in the native compiler"
}

run_command attribute.register `ir_def
run_command attribute.register `ir_decl
run_command attribute.register `compiler_plugin
