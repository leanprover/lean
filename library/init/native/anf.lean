prelude

import init.meta.format
import init.meta.expr
import init.data.string
import init.category.state

import init.native.internal
import init.native.ir
import init.native.format
import init.native.builtin
import init.native.util
import init.native.pass
import init.native.config

@[reducible] meta def binding :=
  (name × expr × expr)

@[reducible] meta def anf_state :=
  (list (list binding) × nat)

@[reducible] meta def anf_monad :=
  state anf_state

meta def trace_anf (s : string) : anf_monad unit :=
  trace s (fun u, return u)

private meta def let_bind (n : name) (ty : expr) (e : expr) : anf_monad unit := do
  scopes <- state.read,
  match scopes with
  | ([], _) := return ()
  | ((s :: ss), count) := state.write $ (((n, ty, e) :: s) :: ss, count)
  end

private meta def mk_let : list binding -> expr -> expr
| [] body := body
| ((n, ty, val) :: es) body :=
  mk_let es (expr.elet n ty val (expr.abstract body (mk_local n)))

private meta def mk_let_in_current_scope (body : expr) : anf_monad expr := do
  (scopes, _) <- state.read,
  match scopes with
  | [] := pure $ body
  | (top :: _) := return $ mk_let top body
  end

private meta def enter_scope (action : anf_monad expr) : anf_monad expr := do
  (scopes, count) <- state.read,
  state.write ([] :: scopes, count),
  result <- action,
  bound_result <- mk_let_in_current_scope result,
  state.write (scopes, count),
  return bound_result

private meta def fresh_name : anf_monad name := do
  (ss, count) <- state.read,
  -- need to replace this with unique prefix as per our earlier conversation
  n <- pure $ name.mk_numeral (unsigned.of_nat count) `_anf_,
  state.write (ss, count + 1),
  return n

-- Hoist a set of expressions above the result of the callback
-- function.
meta def hoist
  (anf : expr -> anf_monad expr)
  (kont : list name -> anf_monad expr) : list expr → anf_monad expr
| [] := kont []
| es := do
     ns <- monad.for es $ (fun x, do
       value <- anf x,
       fresh <- fresh_name,
       let_bind fresh mk_neutral_expr value,
       return fresh),
     kont ns

private meta def anf_constructor (head : expr) (args : list expr) (anf : expr -> anf_monad expr) : anf_monad expr :=
  hoist anf (fun args', return $ mk_call head (list.map mk_local args')) args

private meta def anf_call (head : expr) (args : list expr) (anf : expr -> anf_monad expr) : anf_monad expr := do
  hoist anf (fun ns, match ns with
  -- need to think about how to refactor this, we should get at least one back from here always
  -- this case should never happen
  | [] := return head
  | (head' :: args') := return $ mk_call (mk_local head') (list.map mk_local args')
  end) (head :: args)

private meta def anf_case (action : expr -> anf_monad expr) (e : expr) : anf_monad expr := do
  under_lambda fresh_name (fun e', enter_scope (action e')) e

private meta def anf_cases_on (head : expr) (args : list expr) (anf : expr -> anf_monad expr) : anf_monad expr := do
  match args with
  | [] := return $ mk_call head []
  | (scrut :: cases) := do
    scrut' <- anf scrut,
    cases' <- monad.mapm (anf_case anf) cases,
    return $ mk_call head (scrut' :: cases')
  end

-- stop deleting this, not sure why I keep removing this line of code
open application_kind

private meta def anf' : expr -> anf_monad expr
| (expr.elet n ty val body) := do
  fresh <- fresh_name,
  val' <- anf' val,
  let_bind fresh ty val',
  anf' (expr.instantiate_vars body [mk_local fresh])
| (expr.app f arg) := do
  let fn := expr.get_app_fn (expr.app f arg),
      args := expr.get_app_args (expr.app f arg)
   in match app_kind fn with
   | cases := anf_cases_on fn args anf'
   | constructor := anf_constructor fn args anf'
   | other := anf_call fn args anf'
   end
| e := return e

private meta def init_state : anf_state :=
  ([], 0)

private meta def anf_transform (conf : config) (e : expr) : expr :=
  prod.fst $ (under_lambda fresh_name (enter_scope ∘ anf') e) init_state

meta def anf : pass := {
  name := "anf",
  transform := fun conf proc, procedure.map_body (fun e, anf_transform conf e) proc
}
