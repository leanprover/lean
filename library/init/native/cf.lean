prelude

import init.meta.format
import init.meta.expr
import init.data.string
import init.category.state

import init.native.ir
import init.native.format
import init.native.builtin
import init.native.util
import init.native.pass
import init.native.procedure
import init.native.internal
import init.native.config

namespace cf

open native

@[reducible] meta def cf_state :=
  config Ã— nat

@[reducible] meta def cf_monad :=
  state cf_state

meta def when_debug (action : cf_monad unit) : cf_monad unit := do
  (config, _) <- state.read,
  if config.debug config
  then action
  else return ()

-- point at the code where you can't synthesize?
-- the error behavior here seems bad if you replace the unit
-- with `u`
meta def trace_cf (s : string) : cf_monad unit :=
  when_debug (trace s (fun u, return ()))

meta def fresh_name : cf_monad name := do
  (config, count) <- state.read,
  -- need to replace this with unique prefix as per our earlier conversation
  n <- pure $ name.mk_numeral (unsigned.of_nat count) `_anf_,
  state.write (config, count + 1),
  return n

private meta def cf_case (action : expr -> cf_monad expr) (e : expr) : cf_monad expr := do
  under_lambda fresh_name (fun e', action e') e

private meta def cf_cases_on (head : expr) (args : list expr) (cf : expr -> cf_monad expr) : cf_monad expr :=
  match args with
  | [] := return $ mk_call head []
  | (scrut :: cases) := do
    trace_cf "inside cases on",
    cases' <- monad.mapm (cf_case cf) cases,
    return $ mk_call head (scrut :: cases')
  end

meta def cf' : expr -> cf_monad expr
| (expr.elet n ty val body) :=
  expr.elet n ty val <$> (cf' body)
| (expr.app f arg) := do
  trace_cf "processing app",
  let fn := expr.get_app_fn (expr.app f arg),
      args := expr.get_app_args (expr.app f arg)
   in if is_cases_on fn
   then cf_cases_on fn args cf'
   else return (mk_call (expr.const `native_compiler.return []) [(expr.app f arg)])
| e := return $ expr.app (expr.const `native_compiler.return []) e

meta def init_state : config -> cf_state :=
  fun c, (c, 0)

end cf

private meta def cf_transform (conf : config) (e : expr) : expr :=
  prod.fst $ (under_lambda cf.fresh_name cf.cf' e) (cf.init_state conf)

meta def cf : pass := {
  name := "control_flow",
  transform := fun conf proc, procedure.map_body (fun e, cf_transform conf e) proc
}
