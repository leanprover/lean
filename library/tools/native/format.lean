/-
Copyright (c) 2016 Jared Roesch. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Jared Roesch
-/

import init.meta.name
import init.meta.format
import init.function

import tools.native.ir

meta def format_concat : list format → format
| [] := format.nil
| (f :: fs) := f ++ format_concat fs

meta def comma_sep (items : list format) : format :=
format_concat
  (list.intersperse (format.of_string "," ++ format.space) items)

namespace format_cpp

meta def replace (c : char) (replacement : string) : string -> string
| [] := []
| (s :: ss) :=
  if c = s
  then replacement ++ replace ss
  else c :: replace ss

meta def mangle_name (n : name) : format :=
  to_fmt $ replace #"c" "_$single_quote$_" $ name.to_string_with_sep "_" n

private meta def mk_constructor_args : list name → list format
| [] := []
| (n :: ns) := mangle_name n :: mk_constructor_args ns

private meta def mk_constructor
  (arity : nat)
  (fs : list name) : format :=
  "lean::mk_vm_constructor(" ++ to_fmt arity ++ "," ++
  (format.bracket "{" "}" (comma_sep $ mk_constructor_args fs)) ++ ")"

private meta def mk_call (symbol : name) (args : list name) : format :=
  mangle_name symbol ++ (format.bracket "(" ")" (comma_sep $ list.map mangle_name args))

meta def literal : ir.literal → format
| (ir.literal.nat n) := to_fmt "lean::mk_vm_nat(" ++ to_fmt n ++ ")"
| (ir.literal.string s) := to_fmt "NYI"

meta def format_local (n : name) : format :=
  to_fmt (name.to_string_with_sep "_" n)

meta def string_lit (s : string) : format :=
  format.bracket "\"" "\"" (to_fmt s)

meta def block (body : format) : format :=
format.bracket "{" "}" (format.nest 4 (format.line ++ body) ++ format.line)

meta def expr' (action : ir.stmt → format) : ir.expr → format
| (ir.expr.call f xs) := mk_call f xs
| (ir.expr.mk_object n fs) :=
  if n = 0
  -- Over time I should remove these special case functions,
  -- and just use the def language of the IR.
  then to_fmt "lean::mk_vm_simple(0)"
  else mk_constructor n fs
| (ir.expr.global n) :=
  mk_call n []
| (ir.expr.locl n) :=
  mangle_name n
| (ir.expr.lit l) :=
   literal l
-- project really should only work for like fields/primtive arrays, this is a temporary hack
| (ir.expr.project obj n) :=
  "cfield(" ++ (mangle_name obj) ++ ", " ++ (to_fmt n) ++ ")"
| (ir.expr.panic err_msg) :=
  to_fmt "throw std::runtime_error(" ++ string_lit err_msg ++ ");"
| (ir.expr.mk_native_closure n args) :=
"lean::mk_native_closure(*g_env, lean::name({\"" ++ name.to_string_with_sep "\", \"" n ++ "\"})" ++ "," ++
   format.bracket "{" "}" (comma_sep (list.map format_local args)) ++ ")"
 | (ir.expr.invoke n args) :=
 "lean::invoke(" ++ name.to_string_with_sep "_" n ++ ", " ++
 (comma_sep (list.map format_local args)) ++ ")"
 | (ir.expr.uninitialized) := ";"
 | (ir.expr.constructor _ _) := "NYI"
 | (ir.expr.address_of e) := "& " ++ mangle_name e ++ ";"
 | (ir.expr.equals e1 e2) := expr' e1 ++ " == " ++ expr' e2
 | (ir.expr.raw_int n) := repr n
 | (ir.expr.sub e1 e2) :=
   expr' e1 ++ " - " ++ expr' e2

meta def default_case (body : format) : format :=
  to_fmt "default:" ++ block body

meta def case (action : ir.stmt → format) : (nat × ir.stmt) → format
| (n, s) := "case " ++ to_fmt n ++ ":" ++ block (action s ++ format.line ++ "break;" ++ format.line)

meta def cases (action : ir.stmt → format) : list (nat × ir.stmt) → format
| [] := format.nil
| (c :: cs) := case action c ++ cases cs

meta def ty : ir.ty → format
| (ir.ty.object _) := format.of_string "lean::vm_obj "
| (ir.ty.ref T) := ty T ++ format.of_string " const & "
| (ir.ty.mut_ref T) := ty T ++ format.of_string " &"
| (ir.ty.int) := "int "
| (ir.ty.object_buffer) := "lean::buffer<lean::vm_obj> "
| (ir.ty.name n) := to_fmt n ++ format.space
| (ir.ty.base _) := "NYI"

meta def parens (inner : format) : format :=
  format.bracket "(" ")" inner

meta def stmt : ir.stmt → format
| (ir.stmt.e e) := expr' stmt e ++ ";"
| (ir.stmt.return e) :=
  format.of_string "return"  ++
  format.space ++
  expr' stmt e ++ format.of_string ";"
| (ir.stmt.letb n t ir.expr.uninitialized nop) :=
  ty t ++ (mangle_name n) ++ to_fmt ";" ++ format.line
  -- type checking should establish that these two types are equal
| (ir.stmt.letb n t (ir.expr.constructor ty_name args) nop) :=
  -- temporary hack, need to think about how to model this better
  if ty_name = "lean::name"
  then let ctor_args := comma_sep (list.map (string_lit ∘ to_string) args) in
    ty t ++ (mangle_name n) ++ " = lean::name({" ++ ctor_args ++ "})" ++ to_fmt ";" ++ format.line
  else let ctor_args := parens $ comma_sep (list.map mangle_name args) in
       ty t ++ (mangle_name n) ++ ctor_args ++ to_fmt ";" ++ format.line
| (ir.stmt.letb n t v body) :=
  ty t ++ (mangle_name n) ++ (to_fmt " = ") ++ (expr' stmt v) ++ to_fmt ";" ++
  format.line ++ stmt body
| (ir.stmt.switch scrut cs default) :=
  (to_fmt "switch (") ++ (mangle_name scrut) ++ (to_fmt ")") ++
  (block (format.line ++ cases stmt cs ++ default_case (stmt default)))
| ir.stmt.nop := format.of_string ";"
| (ir.stmt.ite cond tbranch fbranch) :=
  "if (" ++ mangle_name cond ++ ") " ++
    block (stmt tbranch) ++
    block (stmt fbranch) ++ format.line
| (ir.stmt.seq cs) :=
  format_concat (list.map (fun c, stmt c ++ format.line) cs)
| (ir.stmt.assign n val) := mangle_name n ++ " = " ++ expr' stmt val ++ ";"

meta def expr := expr' stmt

meta def format_param (param : name × ir.ty) :=
ty (prod.snd param) ++
format.space ++
to_fmt (name.to_string_with_sep "_" (mk_str_name "_$local$_" (name.to_string_with_sep "_" (prod.fst param))))

meta def format_argument_list (tys : list (name × ir.ty)) : format :=
  format.bracket "(" ")" (comma_sep (list.map format_param tys))

-- meta_def format_prototypes ()
meta def defn (d : ir.defn) : format :=
  match d with
  | ir.defn.mk n arg_tys ret_ty body :=
    let body := stmt body in
    (ty ret_ty) ++ format.space ++ (mangle_name n) ++
    (format_argument_list arg_tys) ++ format.space ++
    (format.bracket "{" "}" $ format.nest 4 (format.line ++ body) ++ format.line)
  end

meta def format_lines : list format → format :=
  fun fs, format_concat $ list.intersperse format.line fs

meta def headers : format :=
  format_lines [
    "#include <iostream>",
    "#include \"util/numerics/mpz.h\"",
    "#include \"library/vm/vm_io.h\"",
    "#include \"library/vm/vm.h\"",
    "#include \"library/io_state.h\"",
    "#include \"init/init.h\"",
    -- pretty sure I can remove this
    "static lean::environment * g_env = nullptr;"
  ]

meta def prototype : ir.defn → format
| (ir.defn.mk n params ret_ty _) :=
  ty ret_ty ++ " " ++ mangle_name n ++ format_argument_list params ++ ";" ++ format.line

meta def defn_prototypes (defs : list ir.defn) : format :=
  format_concat $ list.map prototype defs

meta def split_items : list ir.item → (list ir.defn × list ir.decl)
| [] := ([], [])
| (i :: is) :=
  let (defs, decls) := split_items is in
  match i with
  | ir.item.defn d := (d :: defs, decls)
  | ir.item.decl d := (defs, d :: decls)
  end

meta def declaration : ir.decl → format
| (ir.decl.mk n params ret_ty) :=
  ty ret_ty ++ " " ++ mangle_name n ++ format_argument_list params ++ ";" ++ format.line

meta def declarations (decls : list ir.decl) : format :=
  "namespace lean {\n" ++ format_concat (list.map declaration decls) ++ "}"

meta def definitions (defs : list ir.defn) : format :=
  format_lines $ list.map defn defs

meta def program (items : list ir.item) : format :=
  let (defs, decls) := split_items items in
  format_lines [
    headers,
    defn_prototypes defs,
    declarations decls,
    definitions defs
  ]

end format_cpp
