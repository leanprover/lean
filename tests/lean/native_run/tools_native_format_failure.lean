import tools.native.format
import system.io

meta def main : io unit :=
    put_str $ to_string $ format_cpp.stmt (ir.stmt.letb `f ir.ty.int ir.expr.uninitialized ir.stmt.nop)

vm_eval main
