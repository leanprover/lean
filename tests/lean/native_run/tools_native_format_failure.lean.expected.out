int _$lean$_f;
